### DISCUSSION TITLE
use static library of subproject

### DISCUSSION THREAD

**Discussion by andy5995** [mesonbuild/meson#9830] [Q&A] [answered]
In one of my subprobjects, I have

```
buildtarget = both_libraries(
  meson.project_name(),
  src,
  include_directories : inc,
  version : meson.project_version(),
  soversion : '0',
  install: true)

dep_canfigger = declare_dependency(include_directories : inc,
  link_with : buildtarget)
```

In the superproject, I have
```
canfigger_proj = subproject('canfigger')
dep_canfigger = canfigger_proj.get_variable('dep_canfigger')
```

which works, but I want to link to the static library instead of shared. How can I alter the "static" element of the dep_canfigger object returned by `canfigger_proj.get_variable('dep_canfigger')` so it will be 'true'?

In the subproject, I had

```
dep_canfigger_static = declare_dependency(include_directories : inc,
  link_with : buildtarget.get_static_lib())
dep_canfigger_shared = declare_dependency(include_directories : inc,
  link_with : buildtarget.get_shared_lib())
```

then in the superproject, I used

```dep_canfigger_static = canfigger_proj.get_variable('dep_canfigger_static')```

but I'm thinking there's a better way...

## UPDATE:

I've changed 

 ```
 canfigger_proj = subproject('canfigger')
 dep_canfigger = canfigger_proj.get_variable('dep_canfigger')
 ```
to 

```
dep_canfigger = dependency(
  'canfigger',
  fallback : ['canfigger', 'dep_canfigger_static']
  )
```

I still want it to statically link when using the fallback.

Using this in the subproject (canfigger library meson.build)

```
dep_canfigger_static = declare_dependency(include_directories : inc,
  link_with : buildtarget.get_static_lib())
dep_canfigger_shared = declare_dependency(include_directories : inc,
  link_with : buildtarget.get_shared_lib())
```

isn't so terrible, but I just wanted to check if there's a more efficient way of doing it.

**Comment by eli-schwartz** [comment, marked as answer]
Is there any particular reason to use `both_libraries()` to build both static and shared, rather than using `library()` and letting the person building the project choose which one(s) they want by configuring with `-Ddefault_library={static|shared|both}`?

You can also configure it on a per subproject level using `-Dcanfigger:default_library=static` or by passing `default_options: ['default_library=static']` as a kwarg to the subproject/dependency invocation.

(Also you can add "both" as a default_options in the `project()` definition of canfigger.)

**Reply by andy5995** [reply]
> Is there any particular reason to use `both_libraries()` to build both static and shared, rather than using `library()` and letting the person building the project choose which one(s) they want by configuring with `-Ddefault_library={static|shared|both}`?

Actually, no, now that you mention it. I've made that change to the canfigger meson.build.
> 
> You can also configure it on a per subproject level using `-Dcanfigger:default_library=static` or by passing `default_options: ['default_library=static']` as a kwarg to the subproject/dependency invocation.
> 
> (Also you can add "both" as a default_options in the `project()` definition of canfigger.)

Now when I try to init the superproject builddir with `meson builddir -Dcanfigger:default_library=static`, I get this "unknown method" error.

```
...
canfigger| Compiler for C supports arguments -Wformat-overflow=2: YES (cached)
canfigger| Compiler for C supports arguments -DCANFIGGER_VERSION="0.2.0999": YES
canfigger| Program valgrind found: YES (/usr/bin/valgrind)

../subprojects/canfigger/tests/meson.build:15:0: ERROR: Unknown method "get_shared_lib" in object <StaticLibraryHolder c86eecb@@canfigger@sta: libcanfigger.a> of type StaticLibraryHolder.

A full log can be found at /home/andy/src/rmw-project/rmw/builddir/meson-logs/meson-log.txt
...
```

Btw, I'll be more specific about my use-case. The superproject is [rmw](https://github.com/theimpossibleastronaut/rmw), the subproject (the library), [canfigger](https://github.com/andy5995/canfigger/), is also one of my projects. Because canfigger is not in wide-spread use at this time, my goal is for rmw to static link with canfigger, as opposed to shared which would mean users would have to install the shared library along with the rmw binary. That would just be extra files they wouldn't need (but it's ok if they want to change the default to shared)

But if canfigger *is* already installed (it's in the AUR and some OpenSuse distros), I want meson to find it when building rmw and link to the so.

So this is what's in the canfigger meson.build

```
project(
  'canfigger',
  'c',
  version: '0.2.0999',
  meson_version : '>= 0.48.0',
  default_options: [
    'warning_level=3',
    'default_library=both'
    ]
  )
...
src = ('canfigger.c')
buildtarget = library(
  meson.project_name(),
  src,
  include_directories : inc,
  version : meson.project_version(),
  soversion : '0',
  install: true)

dep_canfigger = declare_dependency(include_directories : inc,
  link_with : buildtarget)
...
```

and in the rmw meson.build:

```
...
dep_canfigger = dependency(
  'canfigger',
  fallback : ['canfigger', 'dep_canfigger']
  )
...

**Reply by andy5995** [reply]
> Now when I try to init the superproject builddir, I get this "unknown method" error.
> 

Correction/clarification... when I init with:

`meson builddir -Dcanfigger:default_library=static`

**Reply by eli-schwartz** [reply]
> Now when I try to init the superproject builddir, I get this "unknown method" error.

Makes sense. get_shared_lib is only useful or valid for a both_libraries(), and in fact I question the value of get_shared_lib there either since using the result is basically identical to using the both_libraries object anyway.

You cannot guarantee get_shared_lib() will succeed if -Ddefault_library=static is used, but in that case the user determined to have a preferred link model of linking statically, so it is all good...

**Reply by andy5995** [reply]
But then nothing works...

```
[andy@prometheus rmw]$ cd builddir/
[andy@prometheus builddir]$ ls
config.h  meson-info  meson-logs  meson-private  meson-uninstalled  po  subprojects
[andy@prometheus builddir]$ ninja
ninja: error: loading 'build.ninja': No such file or directory
[andy@prometheus builddir]$ meson compile
Traceback (most recent call last):
  File "/usr/lib/python3.10/site-packages/mesonbuild/mesonmain.py", line 138, in run
    return options.run_func(options)
  File "/usr/lib/python3.10/site-packages/mesonbuild/mcompile.py", line 324, in run
    cdata = coredata.load(options.wd)
  File "/usr/lib/python3.10/site-packages/mesonbuild/coredata.py", line 1030, in load
    with open(filename, 'rb') as f:
FileNotFoundError: [Errno 2] No such file or directory: '/home/andy/src/rmw-project/rmw/builddir/meson-private/coredata.dat'
```

**Reply by eli-schwartz** [reply]
Of course if `meson setup builddir <options>` fails then ninja won't work either.

The error nevertheless makes sense -- `libfoo.get_shared_lib()` assumes you use both_libraries, but you might as well just use `libfoo` directly. Fix that, and a `meson setup` will work.

**Reply by andy5995** [reply]
Ok, I see that I was still using 'buildtarget.get_shared_lib()' in the canfigger/tests/meson build (I had forgotten to change it).

Thanks again, @eli-schwartz
